### ISSUE TITLE
Performance and load testing

### ISSUE THREAD

**Issue by symroe** [DemocracyClub/UK-Polling-Stations#133]
We will need to deal with a large amount​ of traffic.  This is a placeholder for the work needed to carry this out.

I expect there is a load of denormalizing we can do as a first pass.

**Comment by symroe** [comment]
From @chris48s:

> Given there will be a lot of read action in a short space of time on content that won't change very much over that time, aggressive caching sounds like a good approach. Unfortunately this isn't something I have much previous experience with as the high-traffic stuff I've worked on has tended to be write-heavy (and hence not a good candidate for caching) so I can't really advise on appropriate design decisions, but if you already have specific ideas about how you would want to accomplish this then I might be able to help out with implementation a bit?

**Comment by dhilton** [comment]
I think there are three strands here:

1. Aggressive caching of existing page results 

Given the dataset, the likely hood of your polling station changing in three days is unlikely, thus using something like cloudflare with page rules set to cache the actual html result as well as static content would cover repeat traffic, for part of the polling station lookup. It would need assessing as to the redirect flow about how the caching would work - I think there is a redirect in the current flow that might not be cache friendly?

2. Database / Query optimisation

The postcode lookup makes a number of db lookup including the geospatial lookups, these probably could be merged into less queries, dropping down to raw psql in places to get speed ups. The problem or challenge I guess is that very little is shared between user sessions - an individual postcode mostly covers at most a block of flats[citation needed] which means caching any of the queries in a flow is of minimal value, over doing higher order full page caches. Which means we're into psql query optimisation and adding indexes as sensibly as possible. 

3. Somehow refactor the MapIt call

The geocode function calls MapIt which has a ratelimiter and an API call limit in place. Once the Addressbase work is complete, I _think_ you have the data you need to complete the geocode call without calling out to MapIt. That should be the first optimisation IMHO, as that removes a http request per lookup.

Thoughts? Break this issue into component stories?

**Comment by chris48s** [comment]
closing this as we now have all sorts of multiple server scaling wizardry on the go and PR #651 will mostly remove the need to call mapit. Any further optimisations needed should be raised as a new more specific issue
